### ISSUE TITLE
Implement some kind of squashing outbound buffer entries

### ISSUE THREAD

**Issue by ninja-** [netty/netty#4285]
Following discussion in #4205 we could implement some less hacky way to squash things that has been written, but not yet flushed, maybe per-transport(the outbound buffer entries).
The need for such a thing shows for example when creating a proxy that handles a lot of small packets and then proxes them, that could result in both $Entry + the number of buffers taken with such entries growing the cache to unresonable size.

```
@Scottmitch actually, I finally solved the problem in 100% so I am "out of netty bugs".
The more low-level cumulator wasn't needed. Do you know what FINALLY brings the number of entries from UNLIMITED to just ~ 200? Writing a cumulator that instead of adding a new entry to outbound buffer, writes it to a cummulator buffer and only writes the cummulated buffer when flush() is triggered. With this enabled no stressing in my tests could bring the number of entries to more than ~ 200-300-500 etc.
```
https://gist.github.com/ninja-/91a13b2630210891d429 - MessageSquasher.java - POC working on production

**Comment by ninja-** [comment]
a quick example of a simple imp. at channel outbound buffer:

private ByteBuf internalBuf;
private long pending = 0;
addMessage(msg) {
internalBuf.markWriterIndex();
pending += internalBuf.readableBytes();
internalBuf.writeBytes(msg);
internalBuf.resetWriterIndex();
}

current() { return internalBuf;}

addFlush() {
internalBuf.writerIndex(internalBuf.writerIndex() + pending)
}

**Comment by ninja-** [comment]
before continuing the work on this it's important to explain why this commented out code piece corrupted streams when triggered few times in a row? https://gist.github.com/ninja-/91a13b2630210891d429#file-messagesquasher-java-L116

**Comment by ninja-** [comment]
OK, I think I have implemented this in 10 LOC :) leaving old ideas for reference

and requesting a review of this ChannelOutboundBuffer.java: https://gist.github.com/ninja-/f1d741ea31cd37f1e324

But I couldn't sleep well after deploying this without knowing why the old MessageSquasher fast-path corrupted buffers randomly :(

**Comment by Scottmitch** [comment]
@ninja- - Would you mind submitting a PR? Much easier to review. Thanks!

**Comment by ninja-** [comment]
@Scottmitch yeah working on it, you could see some commits above so I'm nearly done along with a PR for #4284!

**Comment by ninja-** [comment]
done @ PR #4287
